_posts/2020-10-22-how-to-update-to-aws-provider-v3.adoc

```diff
@@ -0,0 +1,191 @@
+:important-caption: :heavy_exclamation_mark:
+:caution-caption: :fire:
+:warning-caption: :warning:
+toc::[]
+endif::[]
+
+== Intro
+
+This guide will walk you through how to update the https://gruntwork.io/reference-architecture/[Gruntwork Reference
+Architecture] and any code that depends on the
+https://gruntwork.io/infrastructure-as-code-library/[Gruntwork Infrastructure as Code Library] to the latest version
+(v3) of the Terraform AWS provider.
+
+=== What you'll learn in this guide
+
+This guide consists of three main sections:
+
+<<core_concepts>>::
+  An overview of version 3 of the AWS provider and why it is important to update your code for compatibility.
+
+<<deployment_walkthrough>>::
+  The steps you need to take to update your code relying on the Gruntwork Infrastructure as Code library and your
+  version of the Gruntwork Reference Architecture with compatibility with AWS provider v3.
+
+
+[[core_concepts]]
+== Core Concepts
+
+=== Background
+
+Version https://github.com/terraform-providers/terraform-provider-aws/releases/tag/v3.0.0[v3.0.0] of the Terraform AWS
+provider was released on July 30th 2020 with backwards incompatible updates. Following the provider release cycle,
+future releases from this point onward would only be compatible with 3.X. This means that if you wish to use any new
+resources or data sources that are released from this point onwards, you will need to be compatible with version 3 of
+the AWS provider to be able to pull those changes in.
+
+Since this is a major version change, there are a number of backwards incompatible changes that were introduced. In
+order to be able to use version 3, you need to update the resources and data sources with backwards incompatible
+changes.
+
+For your own code, you can follow
+https://registry.terraform.io/providers/hashicorp/aws/latest/docs/guides/version-3-upgrade[the official guide] on
+updating to AWS provider version 3. The rest of this guide will focus on updating references to Gruntwork modules and
+the Gruntwork Reference Architecture to be compatible with AWS provider version 3.
+
+
+[[deployment_walkthrough]]
+== Deployment walkthrough
+
+=== Updating references to Gruntwork Infrastructure as Code Library
+
+In order to take advantage of the Terraform AWS provider version 3, you need to update your references to the Gruntwork
+Infrastructure as Code Library to use a compatible version. The following table provides a summary of all the relevant
+Gruntwork AWS modules and the respective versions that are compatible with AWS provider version 3.
+
+[.exceptional]
+IMPORTANT: Gruntwork follows semantic versioning. In practice this means that version updates to the minor version are
+considered backwards incompatible releases for any version updates prior to 1.0.0 release. Make sure to read the release
+notes for the relevant modules any time you are updating minor versions!
+
+[cols="1,1h"]
+|===
+|IaC Package |Minimum version with AWS Provider v3 support
```

Comment:
```suggestion
|Gruntwork Repo |Minimum version with AWS Provider v3 support
```

Reply:
Did we track whether each release required migration steps? Would be great to have a column like, "requires migration steps" and for the ones that require, say, code changes or state changes, we say, "Yes, see migration guide (link)."

Reply:
None of them require actual migrations to the code (the increments are all backwards compatible). However, that is only true if you are going to that version from the immediate previous version. This was why I mentioned the "IMPORTANT" note above.